optax/_src/transform.py

```diff
@@ -344,6 +344,66 @@ def update_fn(updates, state, params=None):
   return base.GradientTransformation(init_fn, update_fn)
 
 
+class ScaleByAmsgradState(NamedTuple):
+  """State for the AMSGrad algorithm."""
+  count: chex.Array  # shape=(), dtype=jnp.int32.
+  mu: base.Updates
+  nu: base.Updates
+  nu_max: base.Updates
+
+
+def scale_by_amsgrad(
+    b1: float = 0.9,
+    b2: float = 0.999,
+    eps: float = 1e-8,
+    eps_root: float = 0.0,
+    mu_dtype: Optional[Any] = None,
+) -> base.GradientTransformation:
+  """Rescale updates according to the AMSGrad algorithm.
+
+  References:
+    [Reddi et al, 2018](https://openreview.net/forum?id=ryQu7f-RZ)
+
+  Args:
+    b1: decay rate for the exponentially weighted average of grads.
+    b2: decay rate for the exponentially weighted average of squared grads.
+    eps: term added to the denominator to improve numerical stability.
+    eps_root: term added to the denominator inside the square-root to improve
+      numerical stability when backpropagating gradients through the rescaling.
+    mu_dtype: optional `dtype` to be used for the first order accumulator; if
+      `None` then the `dtype is inferred from `params` and `updates`.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  mu_dtype = utils.canonicalize_dtype(mu_dtype)
+
+  def init_fn(params):
+    mu = jax.tree_util.tree_map(  # First moment
+        lambda t: jnp.zeros_like(t, dtype=mu_dtype), params)
+    nu = jax.tree_util.tree_map(jnp.zeros_like, params)  # Second moment
+    nu_max = jax.tree_util.tree_map(jnp.zeros_like, params)
+    return ScaleByAmsgradState(count=jnp.zeros([], jnp.int32), mu=mu, nu=nu,
+                               nu_max=nu_max)
+
+  def update_fn(updates, state, params=None):
+    del params
+    mu = update_moment(updates, state.mu, b1, 1)
```

Comment:
Nit: As above, it might be possible to use `scale_by_adam` for parts of this and just set `nu_max` and the `updates` with the special amsgrad code. I personally wouldn't worry about the duplicate calculation of the `updates` since the unused code would be removed by `jax.jit`.

But as above, this is definitely a matter of taste and the advantage of writing it out in full is that it then corresponds to exactly to how it's written in the paper. So I'll leave it up to what you prefer!

Reply:
Actually, let's leave this as is - it's easier to read and doesn't require jit to not do unnecessary calculations.